tgm/nn/model/dygformer.py

```diff
@@ -0,0 +1,585 @@
+"""Adapted from https://github.com/yule-BUAA/DyGLib_TGB."""
+
+from typing import Callable, Tuple
+
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from ..time_encoding import Time2Vec
+
+PADDED_NODE_ID = -1
+
+
+class NeighborCooccurrenceEncoder(nn.Module):
+    r"""An implementation of Neighbor Co-occurrence Encoding Scheme.
+
+    Args:
+        feat_dim (int): dimension of neighbor co-occurrence features (encodings).
+        device (str): Device (cpu or gpu)
+
+    Reference: https://arxiv.org/abs/2303.13047.
+    """
+
+    def __init__(self, feat_dim: int, device: str) -> None:
+        super().__init__()
+        self.feat_dim = feat_dim
+        self.device = device
+
+        self.neighbor_co_occurrence_encoder = nn.Sequential(
+            nn.Linear(in_features=1, out_features=self.feat_dim),
+            nn.ReLU(),
+            nn.Linear(in_features=self.feat_dim, out_features=self.feat_dim),
+        ).to(device)
+
+    # @TODO: Optimize this function to work with torch.Tensor directly and vectorize everything
+    def _count_nodes_freq(
+        self, all_sources_neighbors: np.ndarray, all_dsts_neighbors: np.ndarray
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        assert (
+            all_sources_neighbors.ndim == 2
+            and all_sources_neighbors.shape == all_dsts_neighbors.shape
+        )
+
+        source_freq, dst_freq = [], []
+        for src_neighbors, dst_neighbors in zip(
+            all_sources_neighbors, all_dsts_neighbors
+        ):
+            src_unique_keys, src_inverse_indices, src_counts = np.unique(
+                src_neighbors, return_inverse=True, return_counts=True
+            )
+            # Fequency of each source's neighbor within source's neighbors
+            src_neighbors_freq_src_neighbors = (
+                torch.from_numpy(src_counts[src_inverse_indices])
+                .float()
+                .to(self.device)
+            )
+            src_mapping_dict = dict(zip(src_unique_keys, src_counts))
+
+            dst_unique_keys, dst_inverse_indices, dst_counts = np.unique(
+                dst_neighbors, return_inverse=True, return_counts=True
+            )
+            # Fequency of each destination's neighbor within destination's neighbors
```

Comment:
We can fix these spelling mistakes to avoid getting flagged in later prs, thanks
```suggestion
            # Frequency of each destination's neighbor within destination's neighbors
```